### DISCUSSION TITLE
How does the update logic work?

### DISCUSSION THREAD

**Discussion by gojzdar** [UkrainianBanderasCat/Digital-Logic-Sim2#115] [General]
In most games with logic gates, all gates update at the same time, but here it is different. It doesn't work the same way. For example the usual edge detector ( not(A) AND (A) ) doesn't work here. Can someone explain this? How does the update logic work?

**Comment by just-another-alien** [comment]
Hey @gojzdar , the chip inputs are updated one by one and the outputs are then updated using the result of the operation
i hope that helped!
Kind regards,
JustMindfulYT - Digital Logic Sim 2 Developer

**Comment by gojzdar** [comment]
Hey.
Thank you for taking the time.
I am more curious as to how the changes propagate through the circuit.
I saw some cycle detection scripts, which coincides with the explanation someone else gave me about how he (not Sebastian Lague) implemented his approach, where he picked a starting gate, updated its output value and propagated the change to all connected gates. He mentioned that he only made a quick prototype, so he didn't bother implementing logic that would allow for cycles to function.

I am asking because I also made my own version of digital logic sim. However, I couldn't find any resources about how to simulate the circuit the way Sebastian and you guys did. So I implemented it the first way I thought of - by iterating through all the placed gates, updating their value and storing the new output in a separate variable. After all the gates (NOT, AND, XOR) and chips (user defined components) have been updated, it steps the pins and updates their active value to the value calculated within the tick. My approach acts like all the gates are triggered with a shared clock every tick, resulting in all of them having the same delay.
Because of this, the simple rising edge detector works:

![image](https://github.com/UkrainianBanderasCat/Digital-Logic-Sim2/assets/135328910/00bfbfc9-ef0e-40e9-acbe-8b438e9359cf)

This is how it implemented it as of now. If it isn't too much of a hassle, I'll try to add an option for the simulation to work the same way yours does and for that, I need to know how it is implemented. I am not all that good at reading other people's code, so could you please explain how it works in a little more detail? Thanks.

I haven't created a cross platform version of my program yet and the code is not up to the standard I would want it to have before presenting, so I haven't published it on github yet. 

Thanks in advance.
Gojzdar.

**Reply by just-another-alien** [reply]
Hey there, I cannot really help you, as I did not work on the simulation part myself but you should implement your own way of saving and simulating the chips, as the way we and Sebastian did it is very resource intense as it simulates **every** chip (even chips in chips). I hope that helped but if not you can always ask me or contact me on discord. My discord user is : **justmindfulyt**.
Kind Regards,
JustMindfulYT - Digital Logic Sim 2 Developer

**Reply by gojzdar** [reply]
I already have everything working, including chips and recursive chips. So I only need to work on UI, rendering and saves, beside some QoL features . So you could say it is nearing completion. 
So should I ask the same question on Sebastian's original repo?
Or do you perhaps know someone who worked on the simulation part?

Thank you.

**Reply by just-another-alien** [reply]
Hey there,
Yes I do know someone who worked on the part ( I know and talk with the whole team aka. Lime Inc)
I am gonna tell them to look at your Messages and respond to them! 
kind regards,
JustMindfulYT - Lime Inc

**Reply by UkrainianBanderasCat** [reply]
@gojzdar do you have discord?

**Reply by UkrainianBanderasCat** [reply]
accept my request in discord
